src/quic/endpoint/builder/config.rs

```diff
@@ -152,75 +152,76 @@ impl Config {
 		client_key_pair: Option<KeyPair>,
 		disable_server_verification: bool,
 	) -> ClientConfig {
-		let mut client = ClientConfigBuilder::default();
-
-		// add protocols
-		if !self.protocols.is_empty() {
-			let protocols: Vec<_> = self.protocols.iter().map(Deref::deref).collect();
-			let _ = client.protocols(&protocols);
-		}
-
-		let mut client = client.build();
-
-		// get inner rustls `ClientConfig`
-		{
-			let crypto = Arc::get_mut(&mut client.crypto).expect("failed to build `ClientConfig`");
-
-			// set default root certificates
-			match store {
-				// remove the defaults set by Quinn
-				Store::Empty => {
-					crypto.root_store = RootCertStore::empty();
-					crypto.ct_logs = None;
-				}
-				// is set correctly by Quinn by default
-				Store::Os => (),
-				// insert `webpki-roots`
-				Store::Embedded => {
-					let mut store = RootCertStore::empty();
-					store.add_server_trust_anchors(&webpki_roots::TLS_SERVER_ROOTS);
-					crypto.root_store = store;
+		let mut certificate_store = match store {
+			Store::Empty => RootCertStore::empty(),
+			Store::Os => {
+				let mut store = RootCertStore::empty();
+				if let Ok(certs) = rustls_native_certs::load_native_certs() {
+					for cert in certs {
+						store
+							.add(&rustls::Certificate(cert.0))
+							.expect("invalid native cert");
+					}
 				}
+				store
 			}
-
-			// add custom root certificates
-			let certificates = certificates.into_iter().map(|certificate| {
-				let anchor = trust_anchor_util::cert_der_as_trust_anchor(certificate.as_ref())
-					.expect("`Certificate` couldn't be parsed");
-				OwnedTrustAnchor::from_trust_anchor(&anchor)
-			});
-			crypto.root_store.roots.extend(certificates);
-
-			// insert client certificate
-			if let Some(key_pair) = client_key_pair {
-				crypto.client_auth_cert_resolver = CertificateResolver::new(key_pair);
+			Store::Embedded => {
+				let mut store = RootCertStore::empty();
+				store.add_server_trust_anchors(webpki_roots::TLS_SERVER_ROOTS.0.iter().map(|ta| {
+					OwnedTrustAnchor::from_subject_spki_name_constraints(
+						ta.subject,
+						ta.spki,
+						ta.name_constraints,
+					)
+				}));
+				store
 			}
+		};
+
+		let additional_anchors = certificates.into_iter().map(|certificate| {
+			let anchor = TrustAnchor::try_from_cert_der(certificate.as_ref())
+				.expect("`Certificate` couldn't be parsed");
+			OwnedTrustAnchor::from_subject_spki_name_constraints(
+				anchor.subject,
+				anchor.spki,
+				anchor.name_constraints,
+			)
+		});
+		certificate_store.add_server_trust_anchors(additional_anchors);
+		let mut crypto = rustls::ClientConfig::builder()
+			.with_safe_defaults()
+			.with_root_certificates(certificate_store)
+			.with_no_client_auth();
```

Comment:
This isn't entirely correct, `safe_defaults` allows TLS 1.2, which is incompatible with QUIC.
I already overhauled this function, push coming soon.

Reply:
Someone should really point this out to the quinn authors -- this was taken from their example.